### PR TITLE
VPN-5612 Do not deactivate VPN upon encountering backendfailure

### PR DESCRIPTION
**Pull request by Gela**
## Description

A failed Guardian heartbeat will lead to a back-end failure in the client code, but we should not be ever deactivating the VPN without the user specifically attempting to do so as it can go unnoticed by the user.
This PR removes the call to `backendFailure()` in the connection manager when user is already authenticated if the heartbeat completes with an error code. 

## Reference

    https://mozilla-hub.atlassian.net/browse/VPN-5612

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed

### CODE REVIEW

File: src/connectionmanager.cpp [hunk 1]
```diff
@@ -863,16 +863,15 @@ bool ConnectionManager::switchServers(const ServerData& serverData) {
...
     emit readyToBackendFailure();
     return;
   }
 
   m_nextStep = BackendFailure;
 
-  if (m_state == StateOn || m_state == StateSwitching ||
-      m_state == StateSilentSwitching || m_state == StateConnecting ||
-      m_state == StateCheckSubscription || m_state == StateConfirming) {
+  if (m_state == StateDisconnecting) {
```
Comment (vinocher) [inline src/connectionmanager.cpp:874]: Don't we need to deactivate in the case of real backend failures and health check failures, so we transition to a well defined state? For example backend failure is emitted because of failure to launch the tunnel, handshake with VPN server, any other failures in the daemon. Changing the condition here seems risky because we may end up in an undefined state (activated but not working).

Should we instead stop deactivation only on the Guardian heartbeat failure, because that should not affect the VPN?

File: src/connectionmanager.cpp [hunk 2]
```diff
@@ -872,7 +872,8 @@ void ConnectionManager::backendFailure() {
 
   if (m_state == StateOn || m_state == StateSwitching ||
       m_state == StateSilentSwitching || m_state == StateConnecting ||
-      m_state == StateCheckSubscription || m_state == StateConfirming) {
+      m_state == StateCheckSubscription || m_state == StateConfirming ||
+      m_state == StateDisconnecting) {
```
Comment (vinocher) [inline src/connectionmanager.cpp:876]: Is this change needed?

File: tests/functional/testHeartbeat.js [hunk 3]
```diff
@@ -54,21 +49,30 @@ describe('Backend failure', function() {
...
+    await vpn.forceHeartbeatFailure();
+
+    // TODO: without call to click on cancel this test fails because 
+    // we get stuck on the loading spinner 
+    // after authenticating following the changes in mozillavpn.cpp
+    // to remove the call to backend failure in heartbeatCompleted().
+    // Is clicking on "cancel" here okay?
+    await vpn.waitForQueryAndClick(
+      queries.screenAuthenticating.CANCEL_FOOTER_LINK.visible());
+
```
Comment (vinocher) [inline tests/functional/testHeartbeat.js:61]: Maybe the solution is to raise the backend failure only on StateAuthenticating, because it does make sense to raise a Guardian failure during authentication. So, something like:
void MozillaVPN::heartbeatCompleted(bool success) {
  logger.debug() << "Server-side check done:" << success;

  if (!success) {
    if (state() == StateAuthenticating) {
        m_private->m_connectionManager.backendFailure();
    }
    return;
  }

File: src/tasks/heartbeat/taskheartbeat.cpp [hunk 4]
```diff
@@ -36,12 +36,16 @@ void TaskHeartbeat::run() {
 
             int statusCode = request->statusCode();
 
+            logger.debug() << "Guardian network request encountered an issue. "
```
Comment (vinocher) [inline src/tasks/heartbeat/taskheartbeat.cpp:39]: Couldn't this be merged with the existing logging on line 32 by addind the status code there?

File: src/tasks/heartbeat/taskheartbeat.cpp [hunk 5]
```diff
@@ -36,12 +36,16 @@ void TaskHeartbeat::run() {
...
+                              "HTTP status code: "
+                           << statusCode;
+
             // Internal server errors.
             if (statusCode >= 500 && statusCode <= 509) {
               vpn->heartbeatCompleted(false);
             }
 
-            // Request failure ((?!?)
+            // Client errors.
```
Comment (vinocher) [inline src/tasks/heartbeat/taskheartbeat.cpp:48]: Is the previous comment better?

File: src/mozillavpn.cpp [hunk 6]
```diff
@@ -1216,7 +1216,9 @@ void MozillaVPN::backendServiceRestore() {
 void MozillaVPN::heartbeatCompleted(bool success) {
   logger.debug() << "Server-side check done:" << success;
 
-  if (!success) {
+  // In the event of a Guardian error during authentication,
+  // the "something went wrong" screen is displayed to the user.
```
Comment (vinocher) [inline src/mozillavpn.cpp:1220]: Should the comment explain why (because the service that is used for authentication is unavialble)? Should the comment also say that other heartbeat errors are ignored to prevent unnecessary VPN interruptions?

File: tests/functional/testHeartbeat.js [hunk 7]
```diff
@@ -54,21 +49,26 @@ describe('Backend failure', function() {
     await vpn.waitForQuery(
         queries.screenInitialize.AUTHENTICATE_VIEW.visible());
 
-    await backendFailureAndRestore();
+    // Ensure that in the event of a Guardian error during authentication,
+    // the "something went wrong" screen is displayed to the user.
```
Comment (vinocher) [inline tests/functional/testHeartbeat.js:53]: Super nit: extra white spaces at the end of this line.

File: tests/functional/testHeartbeat.js [hunk 8]
```diff
@@ -122,16 +124,15 @@ describe('Backend failure', function() {
...
 
-      await backendFailureAndRestore();
+      await vpn.forceHeartbeatFailure();
 
       await vpn.waitForQuery(queries.screenHome.CONTROLLER_TITLE.visible());
       assert.equal(
           await vpn.getQueryProperty(
               queries.screenHome.CONTROLLER_TITLE.visible(), 'text'),
-          'VPN is off');
+          'VPN is on');
```
Comment (vinocher) [inline tests/functional/testHeartbeat.js:133]: Should a comment be added here that the VPN should not be disconnected because of heart beat failure?

File: tests/functional/testHeartbeat.js [hunk 9]
```diff
@@ -105,13 +105,15 @@ describe('Backend failure', function() {
...
-      await backendFailureAndRestore();
+      await vpn.forceHeartbeatFailure();
 
       await vpn.waitForQuery(queries.screenHome.CONTROLLER_TITLE.visible());
-      assert.equal(
-          await vpn.getQueryProperty(
-              queries.screenHome.CONTROLLER_TITLE.visible(), 'text'),
-          'VPN is off');
+
+      await vpn.waitForCondition(async () => {
```
Comment (vinocher) [inline tests/functional/testHeartbeat.js:112]: Should a comment be added here that the VPN should not be disconnected because of heart beat failure?

File: src/mozillavpn.cpp (hunk 6 above)
Comment (Gela) [reply src/mozillavpn.cpp:1220]: Definitely worth adding a bit more context!

**Review by lesleyjanenorton — APPROVED** [review]
✨
Thank you @vinocher @Gela !

**Review by vinocher — APPROVED** [review]
